DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/board_test/src/usm_speed.hpp

```diff
@@ -0,0 +1,177 @@
+    // vectors we are checking equality for.
+    long8 compare = (out[i] == reference);
+    for (int j = 0; j < compare.size(); j++) {
+      if (!compare[j]) {
+        std::cerr << "ERROR: Values do not match, "
+                  << "out[" << i << "][" << j << "] = " << out[i][j]
+                  << "; expected " << reference[j] << std::endl;
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+// Parameterized function to perform one of the tests defined above. Allocates
+// and initializes host USM, runs the test, then verifies the answer. Then
+// re-runs the test several times to measure bandwidth.
+int run_test(sycl::queue &q, USMTest test) {
+
+  double kNumBytes_gb;
+
+  // USM host allocation
+  long8 *in = sycl::malloc_host<long8>(kNumItems, q);
+  long8 *out = sycl::malloc_host<long8>(kNumItems, q);
+  if (in == nullptr || out == nullptr) {
+    std::cerr << "Error: Out of memory, can't allocate " << kNumBytes
+              << " bytes" << std::endl;
+    return 1;
+  }
+
+  // Initialize the input with random values and output with zero
+  for (int i = 0; i < kNumItems; i++) {
+    in[i] = rand();
+    out[i] = {0};
+  }
+
+  // Test selection
+  std::function<sycl::event(sycl::queue &, long8 *, long8 *)> kernel;
+  switch (test) {
+  case MEMCOPY:
+    std::cout << "Case: Full Duplex" << std::endl;
+    kernel = memcopy_kernel;
+    // full duplex transfers twice the amount of data
+    kNumBytes_gb = kNumBytes * 2 / kGB;
+    break;
+  case READ:
+    std::cout << "Case: From Host to Device" << std::endl;
+    kernel = read_kernel;
+    kNumBytes_gb = kNumBytes / kGB;
+    break;
+  case WRITE:
+    std::cout << "Case: From Device to Host" << std::endl;
+    kernel = write_kernel;
+    kNumBytes_gb = kNumBytes / kGB;
+    break;
+  default:
+    std::cout << "Error: Failed to launch test" << std::endl;
+    return 1;
+  }
+
+  std::cout << "Iterations: " << kIterations << std::endl;
+  std::cout << "Data size: " << kNumBytes / kMB << " MB" << std::endl;
+  std::cout << "Data type size: " << sizeof(long8) << " bytes" << std::endl;
+
+  // The first iteration is slow due to one time tasks like buffer creation,
+  // program creation and device programming, bandwidth measured in subsequent
+  // iterations.
+  kernel(q, in, out).wait();
+  if (!verify(in, out, test)) {
+    std::cerr << "FAILED" << std::endl << std::endl;
+    return 1;
+  }
+  float time = 0;
+  for (int i = 0; i < kIterations; i++) {
+    sycl::event e = kernel(q, in, out);
+    e.wait();
+    time += SyclGetQStExecTimeNs(e);
+  }
+
+  // Free USM
+  sycl::free(in, q);
+  sycl::free(out, q);
+
+  // Report throughput
+  time /= kIterations;
+  std::cout << "Average Time: " << time / 1000.0 << " ns\t" << std::endl;
+  std::cout << "Average Throughput: "
+            << (kNumBytes_gb / (time / (1000.0 * 1000.0 * 1000.0))) << " GB/s\t"
+            << std::endl << std::endl;
+
+  return 0;
```

Comment:
Very nice adaption from the regtest to this code sample! Looks good to me. Again, One thing is make sure we use the fact that we know what type the data in the array is ahead of time in this code sample. This simplifying assumption changes how the code may be written, so just make sure once the design isn't using old code that can be simplified. Otherwise, looks good!